File: 34. Find First and Last Position of Element in Sorted Array/solution2.cpp
/*
8 ms
13.5 MB

TC: O(Log N)
SC: O(1)

@author: SahilK-027
@breif: Upperbound lowerbound
*/

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
      // To store answer
        vector<int> ans;
        if(binary_search(nums.begin(), nums.end(), target)){
          // If element is present 
          // Return its lower bound index and upperbound index
            ans.push_back(lower_bound(nums.begin(), nums.end(), target) - nums.begin());
          // As upperbound will point to next greater element we have subtracted 1
            ans.push_back(upper_bound(nums.begin(), nums.end(), target) - nums.begin() - 1);
        }
        else{
          // element not found
            ans.push_back(-1);
            ans.push_back(-1);
        }
        return ans;
    }
};

File: 35. Search Insert Position/solution2.js
/*
70ms
42.2 MB

TC: O(log N)
SC: O(1)

@author: SahilK-027
@breif: Binary search

*/
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var searchInsert = function(arr, t) {
    let s = 0;
    let e = arr.length-1;
    let m = parseInt(s + (e-s) / 2);
    console.log(m);
    while(s <= e){
        if(arr[m] === t ){
            return m;
        }
        else if(arr[m] < t){
            s = m + 1;
        }
        else{
            e = m - 1;
        }
        m = parseInt(s + (e-s) / 2);
    }
    return s;
};
